incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSubscribeFactory.java

```diff
@@ -282,52 +346,95 @@ private void onMqttFlush(
             final int reserved = flush.reserved();
 
             assert acknowledge <= sequence;
-            assert sequence >= replySeq;
+            assert sequence >= initialSeq;
+            assert acknowledge >= initialAck;
 
-            replySeq = sequence;
+            initialSeq = sequence;
 
-            assert replyAck <= replySeq;
+            assert initialAck <= initialSeq;
 
             final OctetsFW extension = flush.extension();
             final MqttFlushExFW mqttFlushEx = extension.get(mqttFlushExRO::tryWrap);
 
             assert mqttFlushEx.kind() == MqttFlushExFW.KIND_SUBSCRIBE;
             final MqttSubscribeFlushExFW mqttSubscribeFlushEx = mqttFlushEx.subscribe();
 
-            filters = mqttSubscribeFlushEx.filters();
-            subscriptionIds.clear();
+            currentFilters = mqttSubscribeFlushEx.filters();
+            messagesSubscriptionIds.clear();
+
+            final KafkaFlushExFW kafkaFlushEx = createKafkaFlushEx(currentFilters);
 
+            messages.doKafkaFlush(traceId, authorization, budgetId, reserved, kafkaFlushEx);
 
-            final KafkaFlushExFW kafkaFlushEx =
-                kafkaFlushExRW.wrap(writeBuffer, FlushFW.FIELD_OFFSET_EXTENSION, writeBuffer.capacity())
-                    .typeId(kafkaTypeId)
-                    .merged(m ->
+            if (retainAvailable)
+            {
+                retainedFilters.removeIf(rf -> !currentFilters.anyMatch(f -> f.pattern().asString().equals(rf.filter)));
```

Comment:
Can we avoid the `.asString()` call here?

Reply:
I'm not sure if we can. As we can defer retained filters, we need to remember to retained filters that we have already opened the stream for. So let's say we receive these filters with two mqtt flush: 

1. [<sensor/one>, <sensor/two, retained>]
2. [<sensor/one>, <sensor/two, retained>, <sensor/three, retained>]

In this case for the first mqtt flush, we'll open the sensor/two retained stream. BUT the client subscribed for another filter. We get all the filters again, but we can't open sensor/two stream again. If we use the FW object, we're not "remembering" the ones that were opened already, as it changes constantly.

Reply:
I understand that we might need to hold a reference, just asking if it has to be utf-8 decoded each time or if we can do a more direct comparison, say using StringFW.